Support state transfer in v2MerkleTree::DBAdapter

dartdart26 opened the pull request:
Since state transfer (ST) works in reverse order of block IDs and the
merkle tree supports adding blocks in the opposite (forward) order, add
code to accumulate blocks as they arrive from state transfer and add
them in forward order to the blockchain through the merkle tree. To
implement that, introduce the STTempBlock BFT key subtype to represent a
temporary ST block.

Add a notion of a block being part of the blockchain and of one that is
not yet part of it. For example, the application is only allowed to add
blocks at the end of the blockchain and these blocks are, therefore,
part of the blockchain. State transfer, on the other hand, can add
blocks in arbitrary order and, therefore, these blocks are not part of
the blockchain until all blocks needed to construct a continuous chain
are received. As part of above, discriminate between
DBAdapter::getLastReachableBlock() and DBAdapter::getLatestBlock() and
provide implementations for them. Please see the descriptions in
merkle_tree_db_adapter.h for more information.

When all blocks from ST are successfully accumulated, they can be added
to the blockchain and the temporary keys can be deleted. That needs to
happen atomically and, therefore, a DB transaction is used. In order to
support this flow in memorydb, provide a Transaction class for it.

Since we can properly support adding blocks at the end of the blockchain
only (except for the special case of state transfer), remove the BlockId
parameter from the addBlock(const SetOfKeyValuePairs&) method and make
it automatically add at the end.

Drop support for empty blocks in this PR. It will be provided in a separate
one.

Add unit tests to demonstrate ST-like operation.

dartdart26 commented:
@andrewjstone Thanks for your offline comments! Based on them, I have added a second commit that makes memorydb::Transaction::get() lookup its internal updates first and then, if not found, lookup from storage.

dartdart26 commented:
@andrewjstone Great comments, thanks! I will address them and, additionally, provide a fix for key ordering (as a separate PR) when serializing blocks. The key ordering issue manifests when running read-only replica tests. I will cover that case in the unit test too.

dartdart26 commented on storage/include/blockchain/merkle_tree_db_adapter.h:
```diff
@@ -41,33 +42,61 @@ class DBKeyManipulator : public DBKeyManipulatorBase {
   static sparse_merkle::Hash extractHashFromLeafKey(const concordUtils::Key &key);
 };
 
+// The DBAdapter class provides facilities for managing a key/value blockchain on top of a key/value store.
```
Agreed, it can be made clearer. I will do so when I provide a switch to select direct KV or merkle storage as I will possibly change interfaces and introduce common ones.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -331,16 +391,77 @@ LeafNode DBAdapter::Reader::get_leaf(const LeafKey &key) const {
+    auto block = Sliver{};
+    const auto sTBlockKey = DBKeyManipulator::generateSTTempBlockKey(i);
+    const auto status = db_->get(sTBlockKey, block);
+    if (status.isNotFound()) {
```
I have lifted the requirement on receiving blocks in reverse order, i.e. they can be added in any order. I have also added a test for it. Here, we handle the case of having a chain from `getLastReachableBlock()` to a block that is less than `getLatestBlock()`. If we don't need that, then you are right and we should assert. Please let me know what you think of that.

dartdart26 commented on storage/include/memorydb/transaction.h:
```diff
@@ -0,0 +1,88 @@
+    auto it = updates_.find(key);
+    if (it != std::cend(updates_)) {
+      if (it->second.isDelete) {
+        return std::string{};
```
`concord::storage::rocksdb::Transaction::get()` seems to return when not found:

https://github.com/vmware/concord-bft/blob/master/storage/include/rocksdb/transaction.h#L49

RocksDB documentation is not very clear, though... `rocksdb::DB::Get()` will not touch the passed value (empty in our case), but it doesn't say the same for `rocksdb::Transaction::Get()` - it only says the function is similar.

I would say let's leave as is.

dartdart26 commented on storage/include/blockchain/merkle_tree_db_adapter.h:
```diff
@@ -41,33 +42,61 @@ class DBKeyManipulator : public DBKeyManipulatorBase {
   // - calculating and filling in the parent digest.
   // Typically called by the application when adding a new block.
-  Status addBlock(const concordUtils::SetOfKeyValuePairs &kv, BlockId blockId);
+  Status addBlock(const concordUtils::SetOfKeyValuePairs &kv);
```
Good point, doing that.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -331,16 +391,77 @@ LeafNode DBAdapter::Reader::get_leaf(const LeafKey &key) const {
+      const auto msg = "Failed to commit an addBlock() DB transaction";
+      LOG_ERROR(logger_, msg);
+      return Status::GeneralError(msg);
+    }
```
Good idea, doing that.

dartdart26 commented:
@andrewjstone I have pushed commits that fix a number of issues and address your comments. 

Additionally, I have rebased on:

https://github.com/vmware/concord-bft/pull/434
https://github.com/vmware/concord-bft/pull/433

dartdart26 commented:
@andrewjstone I have dropped support for adding empty blocks in the last commit. It is something we started discussing, but I think we need to continue and decide if that is the right way to go. Also, it relates to whether we want to create empty roots.

ittaiab commented:
> @andrewjstone I have dropped support for adding empty blocks in the last commit. It is something we started discussing, but I think we need to continue and decide if that is the right way to go. Also, it relates to whether we want to create empty roots.

Empty blocks may be prevalent when we add filtering, please also coordinate with @toly-kournik

dartdart26 commented:
> > @andrewjstone I have dropped support for adding empty blocks in the last commit. It is something we started discussing, but I think we need to continue and decide if that is the right way to go. Also, it relates to whether we want to create empty roots.
> 
> Empty blocks may be prevalent when we add filtering, please also coordinate with @toly-kournik

Great point, thanks @ittaiab . Do you refer to a case of having separate trees per client (filter)? I have dropped support for adding empty blocks by the application/user (e.g. the execution engine) in storage in this PR. We can treat empty roots in the merkle tree as a separate issue. I am thinking for filtering we will have another API, though I haven’t spent time on the details and I might be wrong. Do you think users (e.g. the execution engine) should be able to create empty blocks in storage prior to filtering? Please correct me if my understanding is wrong.

andrewjstone commented on storage/test/merkleTreeAdapter_test.cpp:
```diff
@@ -944,6 +957,157 @@ TEST_P(db_adapter, no_blocks) {
+      const auto &referenceBlock = referenceBlockchain[j - 1];
+      ASSERT_TRUE(block == referenceBlock);
+      ASSERT_TRUE(block::getData(block) == block::getData(referenceBlock));
+      ASSERT_TRUE(block::getStateHash(block) == block::getStateHash(referenceBlock));
```
Maybe add another check to ensure the parent hash of block is is equal to the state hash of referenceBlock-1.

dartdart26 replied:
Good point. I have added a check if the parent hash of block is equal to the hash of referenceBlock - 1 . Is that the case you refer to? If not, I will add further tests in a separate PR.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -82,6 +82,16 @@ SetOfKeyValuePairs batchToDbUpdates(const Tree::UpdateBatch &batch) {
+  auto ver = Version{};
+  for (const auto &kv : batch.internal_nodes) {
+    if (kv.first.path().empty() && ver < kv.first.version()) {
+      ver = kv.first.version();
```
Good catch! However, I no longer do that in the latest commit and call `Tree::get_version()` instead. It is linked to setting the `root_` at the end of `Tree::update()` from the cache as we discussed offline.

dartdart26 commented:
Addressed comments from @andrewjstone . Support for adding empty blocks will be provided in a subsequent PR.
